daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/PartialNextElementResolver.scala

```diff
@@ -0,0 +1,298 @@
+  final def popTRD(): TermRuntimeData = {
+    trdStack.pop
+  }
+
+  private def error(
+    iter: ResettableIterator[TermRuntimeData],
+    name: String,
+    namespace: String,
+    optTRD: Option[TermRuntimeData]): Maybe[ElementRuntimeData] = {
+    val allTRDs = {
+      iter.reset()
+      iter.toStream.takeWhile {
+        stackTRD => optTRD.map { _ ne stackTRD }.getOrElse(true)
+      }
+    }
+    //
+    // Gather up all possible elements it could be from
+    // all stack entries up to, not including this one.
+    //
+    // We only get here if all the TRDs on stack had nextElementResolvers
+    // which passed (returning Nope) to indicate that the elements they
+    // had as possibilities are all optional.
+    //
+    // If we get here, we're at the end of a complex element, and we
+    // only get here when the current incoming event was a start element
+    //
+    // So we have a candidate start element, but it wasn't a match anything
+    // possible, and we're expecting an end element for the complex type
+    // at this point.
+    //
+    val allPossibles =
+      allTRDs.flatMap {
+        _.partialNextElementResolver.allPossibleNextElements.map {
+          _.namedQName
+        }
+      }
+    One(new UnexpectedElementErrorERD(optTRD, name, namespace, allPossibles))
+  }
+}
+
+/**
+ * The schema compiler computes this for each Term in a Sequence.
+ *
+ * This is for use assembling the Daffodil Infoset from an InfosetInputter.
+ * Note that not all InfosetInputter's have a concept of namespaces (json is a
+ * prime example of this). In order to support this, each nextElement method
+ * has a variable called hasNamespace, which is used to specify whether or not
+ * the InfosetInputter supports namespaces. If it does not support namespaces,
+ * then the namespace parameter should be ignored. This may affect how the next
+ * element is determined, and may result in an UnparseError in some cases.
+ */
+sealed trait PartialNextElementResolver extends Serializable {
+
+  /**
+   * Returns One(ElementRuntimeData) if the arguments are associated with
+   * such definition.
+   *
+   * Returns Nope if there is no such mapping.
+   */
+  def maybeNextElement(name: String, nameSpace: String, hasNamespace: Boolean): Maybe[ElementRuntimeData]
+
+  // TODO: PERFORMANCE: We really should be interning all QNames so that comparison of QNames can be pointer equality
+  // or nearly so. We're going to do tons of lookups in hash tables, which will compute the hash code, find it is equal,
+  // compare the entire namespace string character by character, only to say yes, and yes will be by far the vast
+  // bulk of the lookup results.  Pointer equality would be so much faster....
+  //
+  def maybeNextElement(nqn: NamedQName, hasNamespace: Boolean): Maybe[ElementRuntimeData] =
+    maybeNextElement(nqn.local, if (hasNamespace) nqn.namespace.toStringOrNullIfNoNS else null, hasNamespace)
+
+  def allPossibleNextElements: Seq[ElementRuntimeData]
```

Comment:
Maybe rename from "all" possible to remove "all" prefix. It's not all anymore, it's a partial set. It's just within this lexically-enclosing model group.